[Apple Pay] Support for PKDisbursementRequest on WebKit

codeOfRobin opened the pull request:
#### c40451f6052e2805fb1c6abfb61fa322c67caf5b
<pre>
[Apple Pay] Support for PKDisbursementRequest on WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=267428">https://bugs.webkit.org/show_bug.cgi?id=267428</a>
<a href="https://rdar.apple.com/115776022">rdar://115776022</a>

Reviewed by Alex Christensen.

This patch adds support for `PKDisbursementRequest` and associated classes. Since `PKDisbursementPaymentRequest` is a subclass and not a property, the code paths may sometimes diverge from convention

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https-expected.txt: Added.
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html: Added.
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.h:
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl:
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayMerchantCapability.h.
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.idl: Copied from Source/WebCore/Modules/applepay/ApplePayMerchantCapability.idl.
* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
* Source/WebCore/Modules/applepay/ApplePayLineItem.h:
* Source/WebCore/Modules/applepay/ApplePayLineItem.idl:
* Source/WebCore/Modules/applepay/ApplePayMerchantCapability.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayMerchantCapability.h:
* Source/WebCore/Modules/applepay/ApplePayMerchantCapability.idl:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.h:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.idl:
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::ApplePayDisbursementPaymentRequest::validate const):
(WebCore::ApplePayDisbursementPaymentRequest::convertAndValidate):
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
(WebCore::ApplePaySessionPaymentRequest::disbursementPaymentRequest const):
(WebCore::ApplePaySessionPaymentRequest::setDisbursementPaymentRequest):
* Source/WebCore/Modules/applepay/PaymentSummaryItems.h:
* Source/WebCore/Modules/applepay/cocoa/PaymentSummaryItemsCocoa.mm:
(WebCore::platformDisbursementSummaryItem):
(WebCore::platformInstantFundsOutFeeSummaryItem):
(WebCore::platformSummaryItem):
(WebCore::platformDisbursementSummaryItems):
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.h:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.idl:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::show):
(WebCore::ApplePayPaymentHandler::shippingAddressUpdated):
(WebCore::ApplePayPaymentHandler::shippingOptionUpdated):
(WebCore::ApplePayPaymentHandler::paymentMethodUpdated):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
(WebCore::MockPaymentCoordinator::completeShippingMethodSelection):
(WebCore::MockPaymentCoordinator::completeShippingContactSelection):
(WebCore::MockPaymentCoordinator::completePaymentMethodSelection):
* Source/WebCore/testing/MockPaymentCoordinator.h:
* Source/WebCore/testing/MockPaymentCoordinator.idl:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.mm:
(WebKit::PaymentAuthorizationPresenter::completePaymentMethodSelection):
(WebKit::PaymentAuthorizationPresenter::completeShippingContactSelection):
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/ApplePay/DisbursementPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayMerchantCapability.h.
* Source/WebKit/Shared/ApplePay/cocoa/DisbursementPaymentRequestCocoa.mm: Added.
(WebKit::platformDisbursementPaymentRequest):
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::toPKMerchantCapabilities):
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/ApplePay/mac/WebPaymentCoordinatorProxyMac.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder&lt;WebCore::ApplePaySessionPaymentRequest&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::ApplePaySessionPaymentRequest&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/274702@main">https://commits.webkit.org/274702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dba65e1cec294312f94428562f606cdb265fc9cc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18735 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42291 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35658 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42063 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21704 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16064 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33162 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40330 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15852 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34394 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13692 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13740 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35347 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43568 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33209 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36142 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35681 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39454 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39382 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14571 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11997 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37733 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state, /WebKitGTK/TestGeolocationManager:/webkit/WebKitGeolocationManager/current-position (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16190 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/34625 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46390 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8922 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16238 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9539 "Found 1 new JSC stress test failure: stress/sampling-profiler-microtasks.js.no-cjit-collect-continuously (failure)") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15834 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -522,6 +522,7 @@ then
     else echo "Unsupported/unknown product: ${PRODUCT_NAME}"
     fi
 elif [[ "${WK_PLATFORM_NAME}" == iphoneos ||
+        "${WK_PLATFORM_NAME}" == iphonesimulator ||
```
What does this do? Given the `if [[ "${WK_PLATFORM_NAME}" =~ .*simulator ]]` branch above, I would think that this condition would never be evaluated when we're building for `iphonesimulator`.

codeOfRobin replied:
Turns out this bit of code for sim support should exist elsewhere. Removed.

rreno commented on Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.h:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
```
It's 2024 already! (please check all the new files)

codeOfRobin replied:
Thank you for reminding me 😅 ! Made this change in all added files: `ApplePayDisbursementPaymentRequest.h/idl`, `DisbursementPaymentRequestCocoa.mm`,  and `DisbursementPaymentRequest.h`

achristensen07 commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -698,6 +713,9 @@ ExceptionOr<void> ApplePayPaymentHandler::shippingAddressUpdated(Vector<RefPtr<A
         update.newDeferredPaymentRequest = WTFMove(applePayModifier.deferredPaymentRequest);
+#endif
+#if ENABLE(APPLE_PAY_DISBURSEMENTS)
+            update.newDisbursementPaymentRequest = WTFMove(applePayModifier.disbursementPaymentRequest);
```
indentation is excessive.

achristensen07 commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -745,6 +763,9 @@ ExceptionOr<void> ApplePayPaymentHandler::shippingOptionUpdated()
         update.newDeferredPaymentRequest = WTFMove(applePayModifier.deferredPaymentRequest);
+#endif
+#if ENABLE(APPLE_PAY_DISBURSEMENTS)
+            update.newDisbursementPaymentRequest = WTFMove(applePayModifier.disbursementPaymentRequest);
```
ditto

achristensen07 commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -267,6 +267,23 @@ ExceptionOr<ApplePayDeferredPaymentRequest> ApplePayDeferredPaymentRequest::conv
+    return { };
+}
+
+ExceptionOr<ApplePayDisbursementPaymentRequest> ApplePayDisbursementPaymentRequest::convertAndValidate() && {
```
WebKit style has { on the beginning of the next line.

codeOfRobin replied:
Huh weirdly enough the style checker failed for me with this message at commit time:

```
ERROR: Source/WebCore/Modules/applepay/ApplePaySession.cpp:277:  Boolean expressions that span multiple lines should have their operators on the left side of the line instead of the right side.  [whitespace/operators] [4]
ERROR: Source/WebCore/Modules/applepay/ApplePaySession.cpp:278:  This { should be at the end of the previous line  [whitespace/braces] [4]
Total errors found: 2 in 45 files
```

I wonder if the && at the end tripped it up?

achristensen07 commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -267,6 +267,23 @@ ExceptionOr<ApplePayDeferredPaymentRequest> ApplePayDeferredPaymentRequest::conv
+
+ExceptionOr<void> ApplePayDisbursementPaymentRequest::validate() const
+{
+    return { };
```
This looks suspiciously incomplete.

codeOfRobin replied:
That's intentional! (For now). Currently the `disbursementPaymentRequest` property is an empty struct, so there's nothing really to validate here.

rreno commented on Source/WebCore/Modules/applepay/ApplePaySession.cpp:
```diff
@@ -332,6 +349,15 @@ static ExceptionOr<ApplePaySessionPaymentRequest> convertAndValidate(Document& d
 
+#if ENABLE(APPLE_PAY_DISBURSEMENTS)
+    if (paymentRequest.disbursementPaymentRequest) {
+        auto disbursementPaymentRequest = WTFMove(*paymentRequest.disbursementPaymentRequest).convertAndValidate();
```
Doesn't the && qualifier for convertAndValidate do the move for us here? So the `WTFMove(*paymentRequest.disbursementPaymentRequest)` is redundant?

rreno replied:
Oh no the `&&` is the qualifier for `this` pointer, not the return type. Disregard.

rreno replied:
Actually the more I think about this the less I like it. Why are we moving an empty struct? Moving it isn't doing anything. This would make more sense if the struct had some well-defined but invalid state that a moved-from object would be in. But, this is an empty struct which has no validation logic.

All that is to say all this moving and `foo() &&` stuff is complicated for no real benefit. I suggest either defining an invalid state that a moved-from object is in and check for that in `validate` or remove all this moving alltogether.

rreno commented on Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
```diff
@@ -255,11 +255,25 @@ ExceptionOr<void> ApplePayPaymentHandler::show(Document& document)
+            lineItems.appendVector(applePayModifier->additionalLineItems);
+
+        request.setLineItems(lineItems);
+    } else if (applePayModifier) {
```
WebKit style says one-liner conditional clauses don't need braces.


```
else if (applePayModifier)
    request.setLineItems(applePayModifier->additionalLineItems);
```

rreno commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -996,6 +996,11 @@ struct WebCore::ApplePayDeferredPaymentRequest {
 #endif
 
+#if ENABLE(APPLE_PAY_DISBURSEMENTS)
+struct WebCore::ApplePayDisbursementPaymentRequest {
```
Not sure if this should be empty.

codeOfRobin replied:
Same as above, these will have properties eventually (for now, simply having a `disbursementPaymentRequest` property existing is sufficient to represent a disbursement.

aprotyas commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -1062,6 +1076,9 @@ struct WebCore::ApplePayLineItem {
 #endif
+#if ENABLE(APPLE_PAY_DISBURSEMENTS)
+    std::optional<WebCore::ApplePayLineItem::DisbursementLineItemType> disbursementLineItemType
+#endifclean
```
This might be messing with the serializer generation
```suggestion
#endif
```

codeOfRobin replied:
Cool let's hope this works

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+    new PaymentRequest([method], validPaymentDetails());
+}, "Should have a feature for `disbursementPaymentRequest`.");
+
+let lineItems = {
```
nit, `const`

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+    }
+}
+
+let validLineItems = [ lineItems.total, lineItems.ifo, lineItems.disbursement ]
```
nit, `const`... same of any other variable that is not reassigned globally.

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+
+function modifiersWithData(data, lineItems) {
+    data.additionalLineItems = lineItems
+    return [ { supportedMethods: 'https://apple.com/apple-pay', data } ];
```
Maybe: 

```suggestion
    return [ { supportedMethods: 'https://apple.com/apple-pay', data: { ...data } } ];
```

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+            }
+        }
+        if (initialDisbursementPaymentRequest !== undefined) {
+            detailsInit.modifiers = modifiersWithData({ disbursementPaymentRequest: initialDisbursementPaymentRequest}, validLineItems);
```
nit: line length. Might be good to run this over [prettier.js](https://prettier.io)

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+    return [ { supportedMethods: 'https://apple.com/apple-pay', data } ];
+}
+
+function testValidDisbursementPaymentRequest(description, {initialDisbursementPaymentRequest, expectedDisbursementPaymentRequest}) {
```
This is usually a bit more maintainable, plus makes it clear the properties won't be reassigned:
 
```suggestion
function testValidDisbursementPaymentRequest(description, options) {
   const {initialDisbursementPaymentRequest, expectedDisbursementPaymentRequest} = options;
```

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+        });
+
+        request.addEventListener("shippingaddresschange", (event) => {
+            event.updateWith({ } );
```
nit: stray ws.

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+
+        let request = new PaymentRequest([validPaymentMethod()], detailsInit);
+        request.addEventListener("merchantvalidation", (event) => {
+            event.complete({});
```
for clarity, its might be good to maybe set:

```JS
const NO_DATA = {}; 
``` 

Or something. Just so it's clear when you call `event.complete({})`

Or better:

```JS
const RESOLVED_PROMISE = Promise.resolve({});
```

As `event.complete` and `event.updateWith()` expect a promise anyway.

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+
+        request.addEventListener("paymentmethodchange", (event) => {
+            let detailsUpdate = { };
+            event.updateWith({  });
```
`detailsUpdate` is defined but never used. 

```suggestion
            event.updateWith({  });
```

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+        }
+        else {
+            assert_equals(actualDisbursementPaymentRequest, expectedDisbursementPaymentRequest, "check that the `DisbursementPaymentRequest` matches");
+        }
```
Maybe just: 
```suggestion
        const { disbursementPaymentRequest } = internals.mockPaymentCoordinator;
        const assert = disbursementPaymentRequest ? assert_object_equals : assert_equals;
        assert(disbursementPaymentRequest, expectedDisbursementPaymentRequest, "check that the `DisbursementPaymentRequest` matches");
```

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,95 @@
+
+done();
+
+</script>
```
```suggestion
</script>

```

marcoscaceres reviewed (changes requested):
Couple of suggestions for improving the test.

achristensen07 commented on Source/WebCore/Modules/applepay/cocoa/PaymentSummaryItemsCocoa.mm:
```diff
@@ -150,6 +179,19 @@ static NSCalendarUnit toCalendarUnit(ApplePayRecurringPaymentDateUnit unit)
+        if (PKPaymentSummaryItem *summaryItem = platformSummaryItem(lineItem))
+            [paymentSummaryItems addObject:summaryItem];
+    }
+    return adoptNS([paymentSummaryItems copy]).autorelease();
```
This copy looks unnecessary.  Why not just return paymentSummaryItems?

achristensen07 commented on Source/WebCore/Modules/applepay/ApplePayLineItem.h:
```diff
@@ -61,6 +61,18 @@ struct ApplePayLineItem final {
+
+#if ENABLE(APPLE_PAY_DISBURSEMENTS)
+
+    enum class DisbursementLineItemType : uint8_t {
```
This could use bool as its underlying type because there are only 2 values.

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html:
```diff
@@ -0,0 +1,116 @@
+    test((test) => {
+      const method = validPaymentMethod();
+      method.data.features = ["paymentRequestDisbursements"];
+      new PaymentRequest([method], validPaymentDetails());
```
Is the expectation here that doesn't throw? (i.e., it's not clear what this is testing)

marcoscaceres reviewed (approved):
Looks great! just that one nit about what's being tested. You can probably leave it as is if you want, as it's kiiiinda implied from the context, but it would be good to double check if the will actually throw.

Commit:
[Apple Pay] Support for PKDisbursementRequest on WebKit
https://bugs.webkit.org/show_bug.cgi?id=267428
rdar://115776022

Reviewed by Alex Christensen.

This patch adds support for `PKDisbursementRequest` and associated classes. Since `PKDisbursementPaymentRequest` is a subclass and not a property, the code paths may sometimes diverge from convention

* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https-expected.txt: Added.
* LayoutTests/http/tests/paymentrequest/ApplePayModifier-disbursementPaymentRequest.https.html: Added.
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.h:
* Source/WebCore/Modules/applepay/ApplePayDetailsUpdateBase.idl:
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayMerchantCapability.h.
* Source/WebCore/Modules/applepay/ApplePayDisbursementPaymentRequest.idl: Copied from Source/WebCore/Modules/applepay/ApplePayMerchantCapability.idl.
* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
* Source/WebCore/Modules/applepay/ApplePayLineItem.h:
* Source/WebCore/Modules/applepay/ApplePayLineItem.idl:
* Source/WebCore/Modules/applepay/ApplePayMerchantCapability.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayMerchantCapability.h:
* Source/WebCore/Modules/applepay/ApplePayMerchantCapability.idl:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.h:
* Source/WebCore/Modules/applepay/ApplePayPaymentRequest.idl:
* Source/WebCore/Modules/applepay/ApplePaySession.cpp:
(WebCore::ApplePayDisbursementPaymentRequest::validate const):
(WebCore::ApplePayDisbursementPaymentRequest::convertAndValidate):
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
(WebCore::ApplePaySessionPaymentRequest::disbursementPaymentRequest const):
(WebCore::ApplePaySessionPaymentRequest::setDisbursementPaymentRequest):
* Source/WebCore/Modules/applepay/PaymentSummaryItems.h:
* Source/WebCore/Modules/applepay/cocoa/PaymentSummaryItemsCocoa.mm:
(WebCore::platformDisbursementSummaryItem):
(WebCore::platformInstantFundsOutFeeSummaryItem):
(WebCore::platformSummaryItem):
(WebCore::platformDisbursementSummaryItems):
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.h:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayModifier.idl:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
(WebCore::ApplePayPaymentHandler::show):
(WebCore::ApplePayPaymentHandler::shippingAddressUpdated):
(WebCore::ApplePayPaymentHandler::shippingOptionUpdated):
(WebCore::ApplePayPaymentHandler::paymentMethodUpdated):
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/PassKitSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
(WebCore::MockPaymentCoordinator::completeShippingMethodSelection):
(WebCore::MockPaymentCoordinator::completeShippingContactSelection):
(WebCore::MockPaymentCoordinator::completePaymentMethodSelection):
* Source/WebCore/testing/MockPaymentCoordinator.h:
* Source/WebCore/testing/MockPaymentCoordinator.idl:
* Source/WebKit/Platform/cocoa/PaymentAuthorizationPresenter.mm:
(WebKit::PaymentAuthorizationPresenter::completePaymentMethodSelection):
(WebKit::PaymentAuthorizationPresenter::completeShippingContactSelection):
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/ApplePay/DisbursementPaymentRequest.h: Copied from Source/WebCore/Modules/applepay/ApplePayMerchantCapability.h.
* Source/WebKit/Shared/ApplePay/cocoa/DisbursementPaymentRequestCocoa.mm: Added.
(WebKit::platformDisbursementPaymentRequest):
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.h:
* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::toPKMerchantCapabilities):
* Source/WebKit/Shared/ApplePay/ios/WebPaymentCoordinatorProxyIOS.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/ApplePay/mac/WebPaymentCoordinatorProxyMac.mm:
(WebKit::WebPaymentCoordinatorProxy::platformShowPaymentUI):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::ApplePaySessionPaymentRequest>::encode):
(IPC::ArgumentCoder<WebCore::ApplePaySessionPaymentRequest>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/274702@main
